File: scripts/import_partner_csv.py
#!/usr/bin/env python3
"""
Import partner CSV (Resident Advisor export) into MySQL.
Skips matched events and logs audit rows.

Requires:
  python3 -m pip install mysql-connector-python

Env vars:
  DB_HOST, DB_PORT, DB_USER, DB_PASSWORD, DB_NAME
Optional:
  CITY_NAME (default: Manchester)
  CITY_SLUG (default: manchester)
  CITY_REGION (default: Greater Manchester)
  CITY_COUNTRY_CODE (default: GB)
  CITY_TIMEZONE (default: Europe/London)

Usage:
  python3 scripts/import_partner_csv.py /path/to/partner.csv
"""

import csv
import os
import re
from datetime import datetime
from typing import Any, Dict, Optional, Tuple

import mysql.connector

DB_HOST = os.getenv("DB_HOST", "127.0.0.1")
DB_PORT = int(os.getenv("DB_PORT", "3306"))
DB_USER = os.getenv("DB_USER", "root")
DB_PASSWORD = os.getenv("DB_PASSWORD", "")
DB_NAME = os.getenv("DB_NAME", "queercity")

CITY_NAME = os.getenv("CITY_NAME", "Manchester")
CITY_SLUG = os.getenv("CITY_SLUG", "manchester")
CITY_REGION = os.getenv("CITY_REGION", "Greater Manchester")
CITY_COUNTRY_CODE = os.getenv("CITY_COUNTRY_CODE", "GB")
CITY_TIMEZONE = os.getenv("CITY_TIMEZONE", "Europe/London")

PRICE_RE = re.compile(r"\d+(?:\.\d+)?")


def parse_datetime(date_str: str, time_str: str) -> Optional[datetime]:
    date_str = (date_str or "").strip()
    time_str = (time_str or "").strip()
    if not date_str:
        return None
    if not time_str:
        time_str = "00:00:00"
    if len(time_str) == 5:
        time_str = time_str + ":00"
    try:
        return datetime.fromisoformat(f"{date_str}T{time_str}")
    except ValueError:
        return None


def normalize_str(value: Any) -> str:
    if value is None:
        return ""
    return str(value).strip()


def parse_price(value: str) -> Optional[float]:
    if not value:
        return None
    m = PRICE_RE.findall(str(value))
    if not m:
        return None
    try:
        return float(m[0])
    except ValueError:
        return None


def split_tags(value: str):
    if not value:
        return []
    return [t.strip().lower() for t in str(value).split(";") if t.strip()]


class Importer:
    def __init__(self, conn):
        self.conn = conn
        self.cur = conn.cursor()
        self.address_cache = {}
        self.place_cache = {}
        self.tag_cache = {}
        self.city_id = None
        self.run_id = None

    def ensure_city(self):
        self.cur.execute("SELECT id FROM cities WHERE slug=%s", (CITY_SLUG,))
        row = self.cur.fetchone()
        if row:
            self.city_id = row[0]
            return
        self.cur.execute(
            "INSERT INTO cities (name, region, country_code, timezone, slug) VALUES (%s,%s,%s,%s,%s)",
            (CITY_NAME, CITY_REGION, CITY_COUNTRY_CODE, CITY_TIMEZONE, CITY_SLUG),
        )
        self.city_id = self.cur.lastrowid

    def start_run(self, source: str):
        self.cur.execute("INSERT INTO import_runs (source) VALUES (%s)", (source,))
        self.run_id = self.cur.lastrowid

    def log_row(self, source_event_id: str, action: str, reason: str = None):
        self.cur.execute(
            "INSERT INTO import_rows (run_id, source_event_id, action, reason) VALUES (%s,%s,%s,%s)",
            (self.run_id, source_event_id or None, action, reason),
        )

    def upsert_address(self, street, locality, postal, country) -> Optional[int]:
        street = normalize_str(street)
        locality = normalize_str(locality)
        postal = normalize_str(postal)
        country = normalize_str(country)
        key = (street, locality, postal, country)
        if key in self.address_cache:
            return self.address_cache[key]
        self.cur.execute(
            "SELECT id FROM postal_addresses WHERE street_address=%s AND address_locality=%s AND postal_code=%s AND address_country=%s",
            (street or None, locality or None, postal or None, country or None),
        )
        row = self.cur.fetchone()
        if row:
            self.address_cache[key] = row[0]
            return row[0]
        self.cur.execute(
            "INSERT INTO postal_addresses (street_address, address_locality, postal_code, address_country) VALUES (%s,%s,%s,%s)",
            (street or None, locality or None, postal or None, country or None),
        )
        addr_id = self.cur.lastrowid
        self.address_cache[key] = addr_id
        return addr_id

    def upsert_place(self, name, street, postal) -> Optional[int]:
        name = normalize_str(name)
        addr_id = self.upsert_address(street, CITY_NAME, postal, CITY_COUNTRY_CODE)
        key = (name, addr_id)
        if key in self.place_cache:
            return self.place_cache[key]
        self.cur.execute(
            "SELECT id FROM places WHERE name=%s AND address_id <=> %s",
            (name or None, addr_id),
        )
        row = self.cur.fetchone()
        if row:
            self.place_cache[key] = row[0]
            return row[0]
        self.cur.execute(
            "INSERT INTO places (name, address_id, city_id) VALUES (%s,%s,%s)",
            (name or None, addr_id, self.city_id),
        )
        place_id = self.cur.lastrowid
        self.place_cache[key] = place_id
        return place_id

    def find_event(self, event_id: str, url: str, name: str, start_dt: Optional[datetime], place_id: Optional[int]) -> Optional[int]:
        if event_id:
            self.cur.execute("SELECT id FROM events WHERE identifier=%s", (event_id,))
            row = self.cur.fetchone()
            if row:
                return row[0]
        if url:
            self.cur.execute("SELECT id FROM events WHERE url=%s", (url,))
            row = self.cur.fetchone()
            if row:
                return row[0]
        if name and start_dt and place_id:
            self.cur.execute(
                "SELECT id FROM events WHERE name=%s AND start_datetime=%s AND place_id=%s",
                (name, start_dt, place_id),
            )
            row = self.cur.fetchone()
            if row:
                return row[0]
        return None

    def insert_event(self, row: Dict[str, str], place_id: Optional[int]) -> int:
        start_dt = parse_datetime(row.get("start_date"), row.get("start_time"))
        end_dt = parse_datetime(row.get("end_date"), row.get("end_time"))
        self.cur.execute(
            """
            INSERT INTO events (
              identifier, name, description, url, image_url, genre, keywords_text,
              event_status, attendance_mode, place_id, city_id, start_datetime, end_datetime
            ) VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)
            """,
            (
                normalize_str(row.get("event_id")) or None,
                normalize_str(row.get("name")),
                row.get("description") or None,
                row.get("url") or None,
                row.get("image") or None,
                row.get("category") or None,
                row.get("tags") or None,
                None,
                None,
                place_id,
                self.city_id,
                start_dt,
                end_dt,
            ),
        )
        return self.cur.lastrowid

    def insert_offer(self, event_id: int, price_value: Optional[float]):
        if price_value is None:
            return
        self.cur.execute(
            "INSERT INTO offers (event_id, price, price_currency, url) VALUES (%s,%s,%s,%s)",
            (event_id, price_value, None, None),
        )

    def insert_tags(self, event_id: int, tags_value: str):
        for tag in split_tags(tags_value):
            if tag in self.tag_cache:
                tag_id = self.tag_cache[tag]
            else:
                self.cur.execute("SELECT id FROM tags WHERE name=%s", (tag,))
                row = self.cur.fetchone()
                if row:
                    tag_id = row[0]
                else:
                    self.cur.execute("INSERT INTO tags (name) VALUES (%s)", (tag,))
                    tag_id = self.cur.lastrowid
                self.tag_cache[tag] = tag_id
            self.cur.execute(
                "INSERT IGNORE INTO event_tags (event_id, tag_id) VALUES (%s,%s)",
                (event_id, tag_id),
            )


def main():
    import sys
    if len(sys.argv) != 2:
        print("Usage: python3 scripts/import_partner_csv.py /path/to/partner.csv")
        sys.exit(1)

    csv_path = sys.argv[1]
    conn = mysql.connector.connect(
        host=DB_HOST,
        port=DB_PORT,
        user=DB_USER,
        password=DB_PASSWORD,
        database=DB_NAME,
        autocommit=False,
    )
    importer = Importer(conn)

    inserted = 0
    skipped = 0
    errors = 0

    try:
        importer.ensure_city()
        importer.start_run(source=os.path.basename(csv_path))

        with open(csv_path, "r", encoding="utf-8") as f:
            reader = csv.DictReader(f)
            for row in reader:
                try:
                    event_id = normalize_str(row.get("event_id"))
                    url = normalize_str(row.get("url"))
                    name = normalize_str(row.get("name"))
                    start_dt = parse_datetime(row.get("start_date"), row.get("start_time"))

                    place_id = importer.upsert_place(
                        row.get("loc_name"),
                        row.get("loc_street"),
                        row.get("loc_post"),
                    )

                    existing_id = importer.find_event(event_id, url, name, start_dt, place_id)
                    if existing_id:
                        importer.log_row(event_id, "skip", "matched existing event")
                        skipped += 1
                        continue

                    new_id = importer.insert_event(row, place_id)
                    importer.insert_offer(new_id, parse_price(row.get("price")))
                    importer.insert_tags(new_id, row.get("tags") or "")
                    importer.log_row(event_id, "insert", "inserted new event")
                    inserted += 1

                except Exception as e:
                    importer.log_row(row.get("event_id"), "error", str(e)[:250])
                    errors += 1

        conn.commit()
        print(f"Import complete: inserted={inserted} skipped={skipped} errors={errors}")
    except Exception:
        conn.rollback()
        raise
    finally:
        conn.close()


if __name__ == "__main__":
    main()

File: scripts/import_partner_csv.php
<?php
// Import partner CSV from a directory (expects exactly one .csv file) into MySQL.
// Skips matched events and logs audit rows. Deletes CSV after successful import.
// Upload this script into: harrymanion.co.uk/queercity/csvimport (same folder as the CSV).
// Run it once via browser, then delete this script.

// ====== CONFIG ======
$DB_HOST = 'db5019616289.hosting-data.io';
$DB_PORT = 3306;
$DB_NAME = 'dbs15283861';
$DB_USER = 'dbu4246002';
$DB_PASS = 'rancEb-tuktor-kyfqi4';

$CITY_NAME = 'Manchester';
$CITY_SLUG = 'manchester';
$CITY_REGION = 'Greater Manchester';
$CITY_COUNTRY_CODE = 'GB';
$CITY_TIMEZONE = 'Europe/London';

$CSV_DIR = __DIR__; // directory containing the CSV (and this script)
$DELETE_CSV_AFTER_SUCCESS = true;
$REQUIRED_TOKEN = 'REPLACE_WITH_SECRET_TOKEN';

// ====== HELPERS ======
function iso_to_mysql_datetime($date, $time) {
    $date = trim((string)$date);
    $time = trim((string)$time);
    if ($date === '') return null;
    if ($time === '') $time = '00:00:00';
    if (strlen($time) === 5) $time .= ':00';
    if (!preg_match('/^\d{4}-\d{2}-\d{2}$/', $date)) return null;
    if (!preg_match('/^\d{2}:\d{2}:\d{2}$/', $time)) return null;
    return $date . ' ' . $time;
}

function normalize_str($v) {
    return $v === null ? '' : trim((string)$v);
}

function split_tags($v) {
    if ($v === null) return [];
    $parts = array_map('trim', explode(';', (string)$v));
    $parts = array_filter($parts, fn($x) => $x !== '');
    return array_values(array_map('strtolower', $parts));
}

function parse_price($v) {
    if ($v === null) return null;
    preg_match_all('/\d+(?:\.\d+)?/', (string)$v, $m);
    if (empty($m[0])) return null;
    return (float)$m[0][0];
}

function find_single_csv($dir) {
    $files = glob($dir . '/*.csv');
    if (!$files || count($files) === 0) {
        throw new Exception('No CSV file found in ' . $dir);
    }
    if (count($files) > 1) {
        throw new Exception('More than one CSV file found; please leave only one in the folder.');
    }
    return $files[0];
}

// ====== MAIN ======
$provided = $_GET['token'] ?? $_POST['token'] ?? '';
if ($REQUIRED_TOKEN === 'REPLACE_WITH_SECRET_TOKEN' || $provided !== $REQUIRED_TOKEN) {
    http_response_code(403);
    echo "Forbidden\\n";
    exit(1);
}

$dsn = "mysql:host=$DB_HOST;port=$DB_PORT;dbname=$DB_NAME;charset=utf8mb4";
$pdo = new PDO($dsn, $DB_USER, $DB_PASS, [
    PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
    PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
]);
$pdo->beginTransaction();

try {
    $csv_path = find_single_csv($CSV_DIR);

    // Ensure city
    $stmt = $pdo->prepare("SELECT id FROM cities WHERE slug = ?");
    $stmt->execute([$CITY_SLUG]);
    $row = $stmt->fetch();
    if ($row) {
        $city_id = (int)$row['id'];
    } else {
        $stmt = $pdo->prepare("INSERT INTO cities (name, region, country_code, timezone, slug) VALUES (?,?,?,?,?)");
        $stmt->execute([$CITY_NAME, $CITY_REGION, $CITY_COUNTRY_CODE, $CITY_TIMEZONE, $CITY_SLUG]);
        $city_id = (int)$pdo->lastInsertId();
    }

    // Start audit run
    $stmt = $pdo->prepare("INSERT INTO import_runs (source) VALUES (?)");
    $stmt->execute([basename($csv_path)]);
    $run_id = (int)$pdo->lastInsertId();

    $log_row = $pdo->prepare("INSERT INTO import_rows (run_id, source_event_id, action, reason) VALUES (?,?,?,?)");

    $get_address = $pdo->prepare("SELECT id FROM postal_addresses WHERE street_address <=> ? AND address_locality <=> ? AND postal_code <=> ? AND address_country <=> ?");
    $ins_address = $pdo->prepare("INSERT INTO postal_addresses (street_address, address_locality, postal_code, address_country) VALUES (?,?,?,?)");

    $get_place = $pdo->prepare("SELECT id FROM places WHERE name <=> ? AND address_id <=> ?");
    $ins_place = $pdo->prepare("INSERT INTO places (name, address_id, city_id) VALUES (?,?,?)");

    $upsert_event = $pdo->prepare(
        "INSERT INTO events (identifier,name,description,url,image_url,genre,keywords_text,event_status,attendance_mode,place_id,city_id,start_datetime,end_datetime)
         VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?)"
    );

    $find_event_id = $pdo->prepare("SELECT id FROM events WHERE identifier=?");
    $find_event_url = $pdo->prepare("SELECT id FROM events WHERE url=?");
    $find_event_fuzzy = $pdo->prepare("SELECT id FROM events WHERE name=? AND start_datetime=? AND place_id=?");

    $ins_offer = $pdo->prepare("INSERT INTO offers (event_id, price, price_currency, url) VALUES (?,?,?,?)");

    $get_tag = $pdo->prepare("SELECT id FROM tags WHERE name=?");
    $ins_tag = $pdo->prepare("INSERT INTO tags (name) VALUES (?)");
    $ins_event_tag = $pdo->prepare("INSERT IGNORE INTO event_tags (event_id, tag_id) VALUES (?,?)");

    $address_cache = [];
    $place_cache = [];
    $tag_cache = [];

    $handle = fopen($csv_path, 'r');
    if (!$handle) throw new Exception('Failed to open CSV');

    $header = fgetcsv($handle);
    if (!$header) throw new Exception('CSV header missing');

    $inserted = 0;
    $skipped = 0;
    $errors = 0;

    while (($row = fgetcsv($handle)) !== false) {
        $data = array_combine($header, $row);
        if (!$data) continue;

        try {
            $event_id = normalize_str($data['event_id'] ?? null);
            $url = normalize_str($data['url'] ?? null);
            $name = normalize_str($data['name'] ?? null);
            $start_dt = iso_to_mysql_datetime($data['start_date'] ?? null, $data['start_time'] ?? null);

            // Address + place
            $street = normalize_str($data['loc_street'] ?? null);
            $postal = normalize_str($data['loc_post'] ?? null);
            $addr_key = $street . '|' . $CITY_NAME . '|' . $postal . '|' . $CITY_COUNTRY_CODE;
            if (isset($address_cache[$addr_key])) {
                $address_id = $address_cache[$addr_key];
            } else {
                $get_address->execute([$street ?: null, $CITY_NAME, $postal ?: null, $CITY_COUNTRY_CODE]);
                $arow = $get_address->fetch();
                if ($arow) {
                    $address_id = (int)$arow['id'];
                } else {
                    $ins_address->execute([$street ?: null, $CITY_NAME, $postal ?: null, $CITY_COUNTRY_CODE]);
                    $address_id = (int)$pdo->lastInsertId();
                }
                $address_cache[$addr_key] = $address_id;
            }

            $place_name = normalize_str($data['loc_name'] ?? null);
            $place_key = $place_name . '|' . $address_id;
            if (isset($place_cache[$place_key])) {
                $place_id = $place_cache[$place_key];
            } else {
                $get_place->execute([$place_name ?: null, $address_id]);
                $prow = $get_place->fetch();
                if ($prow) {
                    $place_id = (int)$prow['id'];
                } else {
                    $ins_place->execute([$place_name ?: null, $address_id, $city_id]);
                    $place_id = (int)$pdo->lastInsertId();
                }
                $place_cache[$place_key] = $place_id;
            }

            // Match existing
            $existing_id = null;
            if ($event_id !== '') {
                $find_event_id->execute([$event_id]);
                $erow = $find_event_id->fetch();
                if ($erow) $existing_id = (int)$erow['id'];
            }
            if (!$existing_id && $url !== '') {
                $find_event_url->execute([$url]);
                $erow = $find_event_url->fetch();
                if ($erow) $existing_id = (int)$erow['id'];
            }
            if (!$existing_id && $name !== '' && $start_dt && $place_id) {
                $find_event_fuzzy->execute([$name, $start_dt, $place_id]);
                $erow = $find_event_fuzzy->fetch();
                if ($erow) $existing_id = (int)$erow['id'];
            }

            if ($existing_id) {
                $log_row->execute([$run_id, $event_id ?: null, 'skip', 'matched existing event']);
                $skipped++;
                continue;
            }

            $end_dt = iso_to_mysql_datetime($data['end_date'] ?? null, $data['end_time'] ?? null);
            $upsert_event->execute([
                $event_id ?: null,
                $name,
                $data['description'] ?? null,
                $data['url'] ?? null,
                $data['image'] ?? null,
                $data['category'] ?? null,
                $data['tags'] ?? null,
                null,
                null,
                $place_id,
                $city_id,
                $start_dt,
                $end_dt,
            ]);
            $new_event_id = (int)$pdo->lastInsertId();

            $price_val = parse_price($data['price'] ?? null);
            if ($price_val !== null) {
                $ins_offer->execute([$new_event_id, $price_val, null, null]);
            }

            foreach (split_tags($data['tags'] ?? null) as $tag) {
                if (isset($tag_cache[$tag])) {
                    $tag_id = $tag_cache[$tag];
                } else {
                    $get_tag->execute([$tag]);
                    $trow = $get_tag->fetch();
                    if ($trow) {
                        $tag_id = (int)$trow['id'];
                    } else {
                        $ins_tag->execute([$tag]);
                        $tag_id = (int)$pdo->lastInsertId();
                    }
                    $tag_cache[$tag] = $tag_id;
                }
                $ins_event_tag->execute([$new_event_id, $tag_id]);
            }

            $log_row->execute([$run_id, $event_id ?: null, 'insert', 'inserted new event']);
            $inserted++;
        } catch (Exception $e) {
            $log_row->execute([$run_id, $event_id ?: null, 'error', substr($e->getMessage(), 0, 250)]);
            $errors++;
        }
    }

    fclose($handle);
    $pdo->commit();

    if ($DELETE_CSV_AFTER_SUCCESS) {
        unlink($csv_path);
    }

    echo "Import complete: inserted=$inserted skipped=$skipped errors=$errors\n";
} catch (Exception $e) {
    $pdo->rollBack();
    echo "ERROR: " . $e->getMessage() . "\n";
    http_response_code(500);
    exit(1);
}
